src/Microsoft.AspNet.Mvc.Core/Routing/AttributeRoute.cs

```diff
@@ -2,275 +2,304 @@
-                }
-                else if (namedEntry == null)
-                {
-                    namedEntries.Add(entry.Name, entry);
-                }
-            }
-
-            _namedEntries = namedEntries;
+        public AttributeRoute(
+            [NotNull] IRouter target,
+            [NotNull] IActionDescriptorsCollectionProvider actionDescriptorsCollectionProvider,
+            [NotNull] IInlineConstraintResolver constraintResolver,
+            [NotNull] ILoggerFactory loggerFactory)
+        {
+            _target = target;
+            _actionDescriptorsCollectionProvider = actionDescriptorsCollectionProvider;
+            _constraintResolver = constraintResolver;
 
-            // The decision tree will take care of ordering for these entries.
-            _linkGenerationTree = new LinkGenerationDecisionTree(linkGenerationEntries.ToArray());
+            _routeLogger = loggerFactory.Create<InnerAttributeRoute>();
+            _constraintLogger = loggerFactory.Create(typeof(RouteConstraintMatcher).FullName);
 
-            _logger = factory.Create<AttributeRoute>();
-            _constraintLogger = factory.Create(typeof(RouteConstraintMatcher).FullName);
+            // Called to create the route eagerly to throw exceptions on startup.
+            GetInnerRoute();
         }
 
         /// <inheritdoc />
-        public async Task RouteAsync([NotNull] RouteContext context)
+        public string GetVirtualPath(VirtualPathContext context)
         {
-            using (_logger.BeginScope("AttributeRoute.RouteAsync"))
-            {
-                foreach (var route in _matchingRoutes)
-                {
-                    var oldRouteData = context.RouteData;
-
-                    var newRouteData = new RouteData(oldRouteData);
-                    newRouteData.Routers.Add(route);
-
-                    try
-                    {
-                        context.RouteData = newRouteData;
-                        await route.RouteAsync(context);
-                    }
-                    finally
-                    {
-                        if (!context.IsHandled)
-                        {
-                            context.RouteData = oldRouteData;
-                        }
-                    }
+            var route = GetInnerRoute();
+            return route.GetVirtualPath(context);
+        }
 
-                    if (context.IsHandled)
-                    {
-                        break;
-                    }
-                }
-            }
+        /// <inheritdoc />
+        public Task RouteAsync(RouteContext context)
+        {
+            var route = GetInnerRoute();
+            return route.RouteAsync(context);
+        }
 
-            if (_logger.IsEnabled(LogLevel.Verbose))
+        private InnerAttributeRoute GetInnerRoute()
+        {
+            var actions = _actionDescriptorsCollectionProvider.ActionDescriptors;
+            if (_inner == null || _inner.Version != actions.Version)
```

Comment:
this is a safe race since inner will never be set to null after it was initially set to not null. Probably worth a comment